# OBTENCIÓ DE DADES

## Agafem dades de l'API opendata de SOM Energia

#### Àngels Villar

#### JULIOL 2021

In [7]:
import numpy as np
import pandas as pd
import requests
import yaml
import os


# SOM ENERGIA API
ENDPOINT_SOMENERGIA = "https://opendata.somenergia.coop/v0.2/"


# List of cities we want to access
cities = ["17114", "17010", "17109", "17133", "17149", "17183", "17162", "17185",
            "17200", "17208", "17139", "17154", "17161", "17165", "17184", "17207", "17021"]

cities_list = []


# List of metrics  we want to access
metrics = ["members", "newmembers", "canceledmembers", "contracts", "newcontracts", "canceledcontracts", "selfconsumptioncontracts",
            "newselfconsumptioncontracts", "canceledselfconsumptioncontracts"]

metrics_list = []



In [8]:
# Construeixo parametres per consultar totes les ciutats de la Garrotxa
params='/by/city/monthly/from/2010-01-01?'
for c in cities:
    params+='&city='+c

In [9]:
# Per cada métrica, get dades aPI 
metrics_list = []
for metric in metrics:
    response = requests.get(ENDPOINT_SOMENERGIA + metric + params)
    print("fet response")
    dades = yaml.safe_load(response.text)
    metrics_list.append(dades['values'])

fet response
fet response
fet response
fet response
fet response
fet response
fet response
fet response
fet response


Converteixo a datagframe per facilitar les tasques d'anàlisi de dades i visualització. També millora eficiència.

In [10]:
dates=dades['dates']
df = pd.DataFrame(list(zip(dates, metrics_list[0],metrics_list[1],metrics_list[2],metrics_list[3],metrics_list[4],metrics_list[5],metrics_list[6],metrics_list[7],metrics_list[8])),
                 columns =["Dates","members", "newmembers", "canceledmembers", "contracts", "newcontracts", "canceledcontracts", "selfconsumptioncontracts",
            "newselfconsumptioncontracts", "canceledselfconsumptioncontracts"])
df

Dates  members  newmembers  canceledmembers  contracts  \
0    2010-01-01        0           0                0          0   
1    2010-02-01        0           0                0          0   
2    2010-03-01        0           0                0          0   
3    2010-04-01        0           0                0          0   
4    2010-05-01        0           0                0          0   
..          ...      ...         ...              ...        ...   
134  2021-03-01      362           7                0        553   
135  2021-04-01      366           5                1        561   
136  2021-05-01      371           5                0        571   
137  2021-06-01      374           3                0        576   
138  2021-07-01      377           3                0        581   

     newcontracts  canceledcontracts  selfconsumptioncontracts  \
0               0                  0                         0   
1               0                  0                         0   
2               0                  0                         0   
3               0                  0                         0   
4               0                  0                         0   
..            ...                ...                       ...   
134            10                  2                        30   
135            14                  7                        31   
136            11                  1                        31   
137             8                  3                        31   
138            13                  8                        32   

     newselfconsumptioncontracts  canceledselfconsumptioncontracts  
0                              0                                 0  
1                              0                                 0  
2                              0                                 0  
3                              0                                 0  
4                              0                                 0  
..                           ...                               ...  
134                            3                                 0  
135                            1                                 0  
136                            0                                 0  
137                            0                                 0  
138                            1                                 0  

[139 rows x 10 columns]

# Guardo el dataframe en _pickle_


Ara que ja tinc les dades, les serialitzo en un fitxer per no haver de tornar a realitzar consultes a l'API.

**Función save_pickle**

Guarda el dataset en formato pickle.

Argumentos: - directorio_datos: directorio donde se guardan ficheros - pickle_file: fichero donde guardaremos el dataset - midataframe: el dataframe a guardar, por defecto es df_ksprojects


In [11]:
DATASETS_PATH = "."
datos_pickle = "somenergia_data.pickle"

def save_pickle(dir_data=DATASETS_PATH,pickle_file=datos_pickle,midataframe=df):

    # Si no existe el directorio, salimos con error
    if not os.path.isdir(dir_data):
        print(directorio_datos,"does not exist!!")
        return print("Error!")

    # el fichero path completo
    picklefile_path = os.path.join(dir_data, pickle_file)

    # abro fichero zip
    midataframe.to_pickle(picklefile_path)
    return picklefile_path



In [12]:
save_pickle()

'.\\somenergia_data.pickle'

In [13]:
df


Dates  members  newmembers  canceledmembers  contracts  \
0    2010-01-01        0           0                0          0   
1    2010-02-01        0           0                0          0   
2    2010-03-01        0           0                0          0   
3    2010-04-01        0           0                0          0   
4    2010-05-01        0           0                0          0   
..          ...      ...         ...              ...        ...   
134  2021-03-01      362           7                0        553   
135  2021-04-01      366           5                1        561   
136  2021-05-01      371           5                0        571   
137  2021-06-01      374           3                0        576   
138  2021-07-01      377           3                0        581   

     newcontracts  canceledcontracts  selfconsumptioncontracts  \
0               0                  0                         0   
1               0                  0                         0   
2               0                  0                         0   
3               0                  0                         0   
4               0                  0                         0   
..            ...                ...                       ...   
134            10                  2                        30   
135            14                  7                        31   
136            11                  1                        31   
137             8                  3                        31   
138            13                  8                        32   

     newselfconsumptioncontracts  canceledselfconsumptioncontracts  
0                              0                                 0  
1                              0                                 0  
2                              0                                 0  
3                              0                                 0  
4                              0                                 0  
..                           ...                               ...  
134                            3                                 0  
135                            1                                 0  
136                            0                                 0  
137                            0                                 0  
138                            1                                 0  

[139 rows x 10 columns]

# EDA
## Descripció de dades
Començo a analitzar les dades que acabo d'obtenir.


In [17]:
df.tail(5)

Dates  members  newmembers  canceledmembers  contracts  \
134  2021-03-01      362           7                0        553   
135  2021-04-01      366           5                1        561   
136  2021-05-01      371           5                0        571   
137  2021-06-01      374           3                0        576   
138  2021-07-01      377           3                0        581   

     newcontracts  canceledcontracts  selfconsumptioncontracts  \
134            10                  2                        30   
135            14                  7                        31   
136            11                  1                        31   
137             8                  3                        31   
138            13                  8                        32   

     newselfconsumptioncontracts  canceledselfconsumptioncontracts  
134                            3                                 0  
135                            1                                 0  
136                            0                                 0  
137                            0                                 0  
138                            1                                 0

A simple vista sense fer massa exploració de les dades, es veu alguna cosa extranya:


fila 136: 371 membres, afegeix 5, cap baixa i el següent valor


fila 137: 374?!

Mirem si ha valors nulls:

In [14]:
df.isnull().sum()

Dates                               0
members                             0
newmembers                          0
canceledmembers                     0
contracts                           0
newcontracts                        0
canceledcontracts                   0
selfconsumptioncontracts            0
newselfconsumptioncontracts         0
canceledselfconsumptioncontracts    0
dtype: int64

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139 entries, 0 to 138
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Dates                             139 non-null    object
 1   members                           139 non-null    int64 
 2   newmembers                        139 non-null    int64 
 3   canceledmembers                   139 non-null    int64 
 4   contracts                         139 non-null    int64 
 5   newcontracts                      139 non-null    int64 
 6   canceledcontracts                 139 non-null    int64 
 7   selfconsumptioncontracts          139 non-null    int64 
 8   newselfconsumptioncontracts       139 non-null    int64 
 9   canceledselfconsumptioncontracts  139 non-null    int64 
dtypes: int64(9), object(1)
memory usage: 11.0+ KB


Els valors semblen correctes, hi ha el manteix número de files en totes les columnes.
Les dates son un objecte, segurament haurem d'arreglar el tipus.


Ara mirarem les principals variables estadístiques:



In [19]:
df.describe()

members  newmembers  canceledmembers   contracts  newcontracts  \
count  139.000000  139.000000       139.000000  139.000000    139.000000   
mean   139.834532    2.769784         0.057554  173.366906      5.280576   
std    113.036905    2.405659         0.233740  182.589194      5.460038   
min      0.000000    0.000000         0.000000    0.000000      0.000000   
25%     40.000000    1.000000         0.000000   14.500000      1.000000   
50%    107.000000    2.000000         0.000000   95.000000      4.000000   
75%    227.500000    4.000000         0.000000  298.500000      9.000000   
max    377.000000   10.000000         1.000000  581.000000     21.000000   

       canceledcontracts  selfconsumptioncontracts  \
count         139.000000                139.000000   
mean            1.129496                  2.791367   
std             1.848754                  7.568503   
min             0.000000                  0.000000   
25%             0.000000                  0.000000   
50%             0.000000                  0.000000   
75%             1.500000                  0.500000   
max             8.000000                 32.000000   

       newselfconsumptioncontracts  canceledselfconsumptioncontracts  
count                   139.000000                        139.000000  
mean                      0.237410                          0.007194  
std                       0.718117                          0.084819  
min                       0.000000                          0.000000  
25%                       0.000000                          0.000000  
50%                       0.000000                          0.000000  
75%                       0.000000                          0.000000  
max                       4.000000                          1.000000

### Análisis de datos

Utilizamos el modulo _ProfileReport_  de la libreria _pandas_profiling_ para generar un fichero con la descripción de los datos del dataframe.

In [21]:
from pandas_profiling import ProfileReport

fichero_profile = "reportprofile.html"

profile = ProfileReport(df, title="Profile")
profile.to_file(fichero_profile)

Summarize dataset:   0%|          | 0/23 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

**[Acces al fitxer creat per pandas_profile](./reportprofile.html)**